### PR TITLE
Added component Validator section

### PR DESCRIPTION
**Pull request by mickaelandrieu**
| Q | A |
| --- | --- |
| Doc fix? | yes |
| New docs? | yes |
| Applies to | >= 2.3 |
| Fixed tickets | I will update this case later |

This pull request is the first from a series that provide a complete documentation for the Validator component.

For now, Installation and Introduction parts will be added, and others sections will refers to the Validator chapters of the Book as suggested by @WouterJ .
- [x] Introduction and Installation of Validator component
- [x] Loading ressources (+ caching)
- [ ] Built-in validators
- [ ] Validation groups
- [ ] Internationalization
- [ ] Custom validation

### CODE REVIEW

**Comment by OskarStark** [comment]
ping @mvhirsch

File: components/validator/builtin_validators.rst [hunk 1]
```diff
@@ -0,0 +1,8 @@
+.. index::
+    single: Validator; Built In validators
+
+Built In validators
```
Comment (OskarStark) [inline components/validator/builtin_validators.rst]: validators -> Validators ?

File: components/validator/custom_validation.rst [hunk 2]
```diff
@@ -0,0 +1,5 @@
+.. index::
+    single: Validator; Custom validation
+
+Custom validation
```
Comment (OskarStark) [inline components/validator/custom_validation.rst]: validation -> Validation?

File: components/validator/builtin_validators.rst (hunk 1 above)
Comment (mickaelandrieu) [inline components/validator/builtin_validators.rst]: Hum ... better to say "Built in validators", what do you think ?

File: components/validator/introduction.rst [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+Installation
+------------
+
+You can install the component in 2 different ways:
+
+* :doc:`Install it via Composer </components/using_components>` (``symfony/validator`` on `Packagist`_);
+* Use the official Git repository (https://github.com/symfony/Validator).
+
+Then, require the vendor/autoload.php file to enable the autoloading mechanism provided by Composer. Otherwise, your application won't be able to find the classes of this Symfony component.
```
Comment (OskarStark) [inline components/validator/introduction.rst]: would add a line break after ```[...] Composer.```

File: components/validator/custom_validation.rst (hunk 2 above)
Comment (mickaelandrieu) [inline components/validator/custom_validation.rst]: Nope, Validation is not a mark|country|person and don't refers to the `Validation` class here.

Comment (OskarStark) [inline components/validator/custom_validation.rst]: :+1: thank you!

File: components/validator/builtin_validators.rst (hunk 1 above)
Comment (WouterJ) [inline components/validator/builtin_validators.rst]: It should be Validators (and below as well). In the docs, we use this title case standard: Capitialize every word except from closed-class words. "Validators/Validation" are not closed class words and have to be capitialized.

File: components/validator/introduction.rst [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+The Validator Component
+=======================
+
+    The Validator component provides tools to validate values following the
+    `JSR-303 Bean Validation specification`_. With the component, this is done in two parts:
+    * ``Contraints``: a constraint describes a rule that need to be validated
+    * ``Validators``: a list of classes that implement the validation logic for common usages
+
+
+
```
Comment (WouterJ) [inline components/validator/introduction.rst]: can you please remove 2 empty lines here?

File: components/validator/introduction.rst (hunk 3 above)
Comment (WouterJ) [inline components/validator/introduction.rst]: Please replace this by:

```rst
.. include:: /components/require_autoload.rst.inc
```

This way, we only need to update one file when something in this paragraph needs to be changed

File: components/validator/introduction.rst [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+:method:`Validation::createValidatorBuilder() <Symfony\\Component\\Validator\\Validation::createValidatorBuilder>`
+method::
+
+    use Symfony\Component\Validator\Validation;
+
+    $validator = Validation::createValidatorBuilder()
+        // ... build a custom instance of the Validator
+        ->getValidator();
+
+What things you can configure will be documented in the following sections.
```
Comment (WouterJ) [inline components/validator/introduction.rst]: I think this is not correct english (I probably wrote it), what about something like: "In the next sections, you'll learn about all things you can configure in the Validator." ?

File: components/validator/resources.rst [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    names. You can also use
+    :method:`Symfony\\Component\\Validator\\ValidatorBuilder::addMethodMappings`
+    to set an array of supported method names.
+
+The FileLoaders
+---------------
+
+The component also provides 2 file loaders, one to load Yaml files and one to
+load XML files. Use 
+:method:`Symfony\\Component\\Validator\\ValidatorBuilder::addYamlMapping` or
```
Comment (WouterJ) [inline components/validator/resources.rst]: we should add a note directive that one has to install symfony/yaml in order to use Yaml Mapping (e.g. "When using Yaml to configure metadata, make sure the ``symfony/yaml`` component is installed in the project: ``composer require symfony/yaml``"

File: components/validator/resources.rst [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+    :method:`Symfony\\Component\\Validator\\ValidatorBuilder::addXmlMappings`
+    to configure an array of file paths.
+
+The AnnotationLoader
+--------------------
+
+At last, the component provides an
+:class:`Symfony\\Component\\Validator\\Mapping\\Loader\\AnnotationLoader`.
+This loader uses an annotation reader to parse the annotations of a class.
+Annotations are placed in doc block comments (`/** ... */`) and start with an
```
Comment (WouterJ) [inline components/validator/resources.rst]: In reStructured text, inline code blocks are marked with double backticks: ``` ``/** ... */`` ```

File: components/validator/resources.rst [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+``@``. For instance::
+
+    // ...
+
+    /**
+     * @Assert\NotBlank()
```
Comment (WouterJ) [inline components/validator/resources.rst]: please also add the use statement before `// ...` in this code block: `use Symfony\Component\Validator\Constraints as Assert;`

File: components/validator/resources.rst [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+    $validator = Validation::createValidatorBuilder()
+        ->addMethodMapping('loadValidatorMetadata')
+        ->getValidator();
+
+Now, the retrieved ``Validator`` tries to find the ``loadValidatorMetadata()``
+method of the class to validate to load its metadata::
+
+    use Symfony\Component\Validator\Mapping\ClassMetadata;
+    use Symfony\Component\Validator\Constraints\NotBlank;
+    use Symfony\Component\Validator\Constraints\Length;
```
Comment (WouterJ) [inline components/validator/resources.rst]: to be consistent with the newer documentation, please replace the 2 constraint use statements by `use Symfony\Component\Validator\Constraints as Assert;` and use `Assert\NotBlank` and `Assert\Length` below

File: components/validator/resources.rst [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+        ->enableAnnotationMapping()
+        ->getValidator();
+
+To disable the annotation loader after it was enabled, call
+:method:`Symfony\\Component\\Validator\\ValidatorBuilder::disableAnnotationMapping`.
+
+.. note::
+
+    In order to use the annotation loader, you should have installed the
+    ``doctrine/annotations`` and ``doctrine/cache`` packages from Packagist.
```
Comment (WouterJ) [inline components/validator/resources.rst]: maybe add the require command here as well: `composer require doctrine/cache doctrine/annotations`

File: components/validator/resources.rst [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+        ->getValidator();
+
+Caching
+-------
+
+Using many loaders to load metadata from different places is very easy when
+creating the metadata, but it can easily slow down your application since each
+file needs to be parsed, validated and converted to a
+:class:`Symfony\\Component\\Validator\\Mapping\\ClassMetadata` instance. To
+solve this problems, you can configure a cacher which will be used to cache
```
Comment (WouterJ) [inline components/validator/resources.rst]: "this problem" (singular)

File: components/validator/resources.rst [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+the ``ClassMetadata`` after it was loaded.
+
+The Validator component comes with a
```
Comment (WouterJ) [inline components/validator/resources.rst]: Should be: "[...] with an" (next word starts with "A")

**Comment by WouterJ** [comment]
Thanks @mickaelandrieu! Do you plan to fill the empty articles in the diff before this PR can be merged? (otherwise, I propose to remove the empty files)

**Comment by mickaelandrieu** [comment]
@WouterJ thanks for review, I will improve my work tonight.

I have question on what did you expect to see in Metadata section ? I will fill the empty articles if we agree to the suggested plan, or I can remove theses to get this first pull request OK: as you like :)

File: components/validator/resources.rst (hunk 6 above)
Comment (mickaelandrieu) [inline components/validator/resources.rst]: agree, I add it  (and we need also a note when we will talk about Form component integration)

File: components/validator/resources.rst (hunk 7 above)
Comment (mickaelandrieu) [inline components/validator/resources.rst]: didn't know, thank you :)

File: components/validator/resources.rst (hunk 10 above)
Comment (mickaelandrieu) [inline components/validator/resources.rst]: hum.. or enrich the Packagist word with a link ?

**Comment by OskarStark** [comment]
when will the first part of this PR be available for review/merge?

**Comment by mickaelandrieu** [comment]
@OskarStark I think I need to remove empty sections and this will be ready to merge.  I'll ping you when it's done :)

**Comment by OskarStark** [comment]
thank you @mickaelandrieu

**Comment by javiereguiluz** [comment]
What can we do to help this get merged? We recently deleted the old README files for all components, so at the moment there are no docs for the Validator component. This is causing troubles for some users (see https://github.com/symfony/symfony/issues/18108).

**Comment by mickaelandrieu** [comment]
@javiereguiluz I'll remove empty sections and rebase against master branch during the weekend. 
Sorry for delay, I've totally forgot this old contribution ^^

**Comment by WouterJ** [comment]
@javiereguiluz any reason why the intro docs were removed? They're very usefull (even for already documented components)...

**Comment by mickaelandrieu** [comment]
@javiereguiluz I don't have time to give some love on this old contribution, sorry. I agree to let someone else finish it.

**Comment by javiereguiluz** [comment]
@mickaelandrieu don't worry! I totally understand. As I said, trying to write the entire documentation for a component is a big mistake. It's better to bootstrap a small introduction and a basic example. Then merge it and improve it. Trying to do everything at once is too much.

**Comment by mickaelandrieu** [comment]
@javiereguiluz I do agree, was entirely my fault to not finish it when I had time to do :/

What can we keep on this contribution to make it merged in your opinion ? If I only had to remove empty section and keep it "as it" it's ok to me, I can do it now.

Tell me :cat2:

**Comment by javiereguiluz** [comment]
@mickaelandrieu thank you very much for working on this. To move things forwards, I've created #6584 to finish your work and do some tweaks. Of course I've reused your original commits, so you'll get full credit for your work.

**Comment by xabbuh** [comment]
Closing in favour of #6584 then. Thanks for kicking this off @mickaelandrieu!

**Pull request closed without merging**